src/main/java/fr/sncf/osrd/infra_state/SwitchPost.java

```diff
@@ -0,0 +1,204 @@
+        var index = currentIndex.get(switchID);
+        var trainID = tables.get(switchID).table.get(index);
+        var count = occurences.get(switchID).get(trainID);
+        occurences.get(switchID).put(trainID, count - 1);
+        currentIndex.put(switchID, index + 1);
+    }
+
+    private void process(Simulation sim, Request request) throws SimulationError {
+
+        var trainID = request.train.schedule.trainID;
+
+        // check if the route is free
+        for (var tvdSectionPath : request.routeState.route.tvdSectionsPaths) {
+            var tvdSectionIndex = tvdSectionPath.tvdSection.index;
+            if (sim.infraState.getTvdSectionState(tvdSectionIndex).isReserved()) {
+                return;
+            }
+        }
+        // check if the train is next on each switch of the route
+        for (var s : request.routeState.route.switchesPosition.keySet()) {
+            if (!isPlanned(s.id, trainID)) { // plan the train if not planned
+                plan(s.id, trainID);
+            }
+            if (!isNext(s.id, trainID)) { // check if next
+                return;
+            }
+        }
+
+        // erase the request of the waiting list of each tvd section of the route
+        for (var tvdSectionPath : request.routeState.route.tvdSectionsPaths) {
+            waitingList.get(tvdSectionPath.tvdSection.id).remove(request);
+        }
+
+        // go to next train to each switch of the route
+        for (var s : request.routeState.route.switchesPosition.keySet()) {
+            next(s.id);
+            currentTrainAllowed.put(s.id, request.train.schedule.trainID);
+        }
+
+        // reserve the route
+        request.routeState.reserve(sim);
+    }
+
+    /**
+     * request a route
+     * 
+     * @param sim        the simulation
+     * @param routeState the route to reserve
+     * @param train      the train that send the request
+     * @throws SimulationError throw a SimulationError if an error happend during process
+     */
+    public void request(Simulation sim, RouteState routeState, Train train) throws SimulationError {
+        var trainID = train.schedule.trainID;
+        if (!lastRequestedRoute.containsKey(trainID) || !lastRequestedRoute.get(trainID).equals(routeState.route.id)) {
+            lastRequestedRoute.put(trainID, routeState.route.id);
+
+            var request = new Request(train, routeState);
+            for (var tvdSectionPath : routeState.route.tvdSectionsPaths) {
+                var tvdSectionID = tvdSectionPath.tvdSection.id;
+                waitingList.get(tvdSectionID).add(request);
+            }
+            process(sim, request);
+        }
+    }
+
+    /**
+     * Notify the SwitchPost that a TVD section is released
+     * @param sim        the simulation
+     * @param tvdSection the TVD section that is released
+     * @throws SimulationError throw a SimulationError if an error happend during process
+     */
+    @SuppressWarnings("unchecked")
```

Comment:
Same remark as in SuccessionTable.java, you can either check the type with an assert or call the copy constructor